lib/using_restricted_array.rb

```diff
@@ -4,60 +4,204 @@
 # All values are integers in the range of 1-221.
 # RestrictedArray cannot be resized.
 
+############################################################################
+
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
+
 def length(array)
-  raise NotImplementedError
+  i = 0
+  while array[i] != nil
+    i += 1
+  end
+  return i
 end
 
+# DELETE AFTER TESTING
+puts "testing length method"
+test_arr = RestrictedArray.new(3)
+pp test_arr
+p length(test_arr)
+puts ""
+
+############################################################################
+
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def print_array(array)
-  raise NotImplementedError
+  i = 0
+  while array[i] != nil
+    puts "#{array[i]}"
+    i += 1
+  end
 end
 
+# DELETE AFTER TESTING
+puts "testing print_arr method"
+test_arr = RestrictedArray.new(3)
+pp test_arr
+p print_array(test_arr)
+puts ""
+
+############################################################################
+
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  if length == 0
+    return nil
+  end
+
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+    end
+  return false
 end
 
+# DELETE AFTER TESTING
+puts "testing search method"
+test_arr = [188, 45, 129]
+pp test_arr
+p search(test_arr, 3, 129)
+puts ""
+
+############################################################################
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n); requires n steps or n operations where n is the length of the array
+# Space complexity: O(n); requires n steps or n operations where n is the length of the array
 def find_largest(array, length)
```

Comment:
👍 However your space complexity is O(1) since you are not building any other data structure.  Your method doesn't increase in memory usage if the input is bigger.